### PR TITLE
s3_object: Add parameter `acl_disabled` to handle uploading files to buckets with ACL disabled.

### LINKED ISSUE #863: aws_s3 can't `put` when bucket has ACLs disabled with `object_ownership` = `BucketOwnerEnforced`
### Summary

When I try to use `aws_s3 ` to upload a file to a bucket where ACLs are disabled, it fails with:

> An exception occurred during task execution. To see the full traceback, use -vvv. The error was: botocore.exceptions.ClientError: An error occurred (AccessControlListNotSupported) when calling the PutObjectAcl operation: The bucket does not allow ACLs

I suspect that the module is simply attempting to specify ACLs for uploaded files unconditionally, and would either need to check whether the bucket has ACLs disabled before hand, or accept object ACLs as configuration items.

### Issue Type

Bug Report

### Component Name

aws_s3

### Ansible Version

```console (paste below)
$ ansible --version`

ansible [core 2.12.4]
  config file = /home/tyler/Repos/px-partners/ansible.cfg
  configured module search path = ['/home/tyler/Repos/px-partners/programmatic-roles/library']
  ansible python module location = /home/tyler/.local/lib/python3.10/site-packages/ansible
  ansible collection location = /home/tyler/.ansible/collections:/usr/share/ansible/collections
  executable location = /home/tyler/.local/bin/ansible
  python version = 3.10.4 (main, Mar 24 2022, 13:07:27) [GCC 11.2.0]
  jinja version = 3.1.1
  libyaml = True


```

### Collection Versions

```console (paste below)
$ ansible-galaxy collection list

# /home/tyler/.local/lib/python3.10/site-packages/ansible_collections
Collection                    Version
----------------------------- -------
amazon.aws                    2.2.0  
ansible.netcommon             2.6.1  
ansible.posix                 1.3.0  
ansible.utils                 2.5.2  
ansible.windows               1.9.0  
arista.eos                    3.1.0  
awx.awx                       19.4.0 
azure.azcollection            1.12.0 
check_point.mgmt              2.3.0  
chocolatey.chocolatey         1.2.0  
cisco.aci                     2.2.0  
cisco.asa                     2.1.0  
cisco.intersight              1.0.18 
cisco.ios                     2.8.1  
cisco.iosxr                   2.9.0  
cisco.ise                     1.2.1  
cisco.meraki                  2.6.1  
cisco.mso                     1.4.0  
cisco.nso                     1.0.3  
cisco.nxos                    2.9.1  
cisco.ucs                     1.8.0  
cloud.common                  2.1.0  
cloudscale_ch.cloud           2.2.1  
community.aws                 2.4.0  
community.azure               1.1.0  
community.ciscosmb            1.0.4  
community.crypto              2.2.4  
community.digitalocean        1.16.0 
community.dns                 2.0.9  
community.docker              2.3.0  
community.fortios             1.0.0  
community.general             4.7.0  
community.google              1.0.0  
community.grafana             1.3.3  
community.hashi_vault         2.4.0  
community.hrobot              1.2.3  
community.kubernetes          2.0.1  
community.kubevirt            1.0.0  
community.libvirt             1.0.2  
community.mongodb             1.3.3  
community.mysql               2.3.5  
community.network             3.1.0  
community.okd                 2.1.0  
community.postgresql          1.7.1  
community.proxysql            1.3.1  
community.rabbitmq            1.1.0  
community.routeros            2.0.0  
community.sap                 1.0.0  
community.skydive             1.0.0  
community.sops                1.2.1  
community.vmware              1.18.0 
community.windows             1.9.0  
community.zabbix              1.5.1  
containers.podman             1.9.3  
cyberark.conjur               1.1.0  
cyberark.pas                  1.0.13 
dellemc.enterprise_sonic      1.1.0  
dellemc.openmanage            4.4.0  
dellemc.os10                  1.1.1  
dellemc.os6                   1.0.7  
dellemc.os9                   1.0.4  
f5networks.f5_modules         1.15.0 
fortinet.fortimanager         2.1.4  
fortinet.fortios              2.1.4  
frr.frr                       1.0.3  
gluster.gluster               1.0.2  
google.cloud                  1.0.2  
hetzner.hcloud                1.6.0  
hpe.nimble                    1.1.4  
ibm.qradar                    1.0.3  
infinidat.infinibox           1.3.3  
infoblox.nios_modules         1.2.1  
inspur.sm                     1.3.0  
junipernetworks.junos         2.10.0 
kubernetes.core               2.3.0  
mellanox.onyx                 1.0.0  
netapp.aws                    21.7.0 
netapp.azure                  21.10.0
netapp.cloudmanager           21.15.0
netapp.elementsw              21.7.0 
netapp.ontap                  21.17.3
netapp.storagegrid            21.10.0
netapp.um_info                21.8.0 
netapp_eseries.santricity     1.3.0  
netbox.netbox                 3.6.0  
ngine_io.cloudstack           2.2.3  
ngine_io.exoscale             1.0.0  
ngine_io.vultr                1.1.1  
openstack.cloud               1.7.2  
openvswitch.openvswitch       2.1.0  
ovirt.ovirt                   1.6.6  
purestorage.flasharray        1.12.1 
purestorage.flashblade        1.9.0  
sensu.sensu_go                1.13.0 
servicenow.servicenow         1.0.6  
splunk.es                     1.0.2  
t_systems_mms.icinga_director 1.28.0 
theforeman.foreman            2.2.0  
vyos.vyos                     2.8.0  
wti.remote                    1.0.3  

# /home/tyler/.ansible/collections/ansible_collections
Collection            Version
--------------------- -------
amazon.aws            3.3.0  


```

### AWS SDK versions

```console (paste below)
$ pip show boto boto3 botocore

WARNING: Package(s) not found: boto
Name: boto3
Version: 1.21.45
Summary: The AWS SDK for Python
Home-page: https://github.com/boto/boto3
Author: Amazon Web Services
Author-email: 
License: Apache License 2.0
Location: /home/tyler/.local/lib/python3.10/site-packages
Requires: botocore, jmespath, s3transfer
Required-by: 
---
Name: botocore
Version: 1.24.45
Summary: Low-level, data-driven core of boto 3.
Home-page: https://github.com/boto/botocore
Author: Amazon Web Services
Author-email: 
License: Apache License 2.0
Location: /home/tyler/.local/lib/python3.10/site-packages
Requires: jmespath, python-dateutil, urllib3
Required-by: boto3, s3transfer


```

### Configuration

```console (paste below)
$ ansible-config dump --only-changed

```

### OS / Environment

_No response_

### Steps to Reproduce

```yaml (paste below)
- name: Upload a file to the bucket
  aws_s3:
    bucket: "{{ s3_bucket_name }}"
    object: "/out/test.txt"
    src: "test.txt"
    mode: put
  changed_when: false

```

### Expected Results

I expect the file to upload without issue.

### Actual Results

```console (paste below)
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: botocore.exceptions.ClientError: An error occurred (AccessControlListNotSupported) when calling the PutObjectAcl operation: The bucket does not allow ACLs
fatal: [localhost]: FAILED! => {"boto3_version": "1.21.45", "botocore_version": "1.24.45", "changed": false, "error": {"code": "AccessControlListNotSupported", "message": "The bucket does not allow ACLs"}, "msg": "Unable to set object ACL: An error occurred (AccessControlListNotSupported) when calling the PutObjectAcl operation: The bucket does not allow ACLs", "response_metadata": {"host_id": "u0/PfRLv4ecD7ZNQnx4VDI7OXbsLSNoeL6icRX1HxoAQ1kbha7eX9YyhmC50b/k/AORoReWhEQ4=", "http_headers": {"connection": "close", "content-type": "application/xml", "date": "Mon, 06 Jun 2022 17:19:35 GMT", "server": "AmazonS3", "transfer-encoding": "chunked", "x-amz-id-2": "u0/PfRLv4ecD7ZNQnx4VDI7OXbsLSNoeL6icRX1HxoAQ1kbha7eX9YyhmC50b/k/AORoReWhEQ4=", "x-amz-request-id": "8PWXR597C2TTN6CX"}, "http_status_code": 400, "request_id": "8PWXR597C2TTN6CX", "retry_attempts": 0}}


```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by mandar242**
##### SUMMARY

Fixes #863 

[Update: 07/07/2022]
https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client.get_bucket_ownership_controls can be used to determine if ACL is disabled or not.
Modified code to convert `acl_disabled` from a user input parameter to a variable used for testing if ACL is enabled/disabled.

Add parameter  `acl_disabled` to handle uploading files to buckets with ACL disabled.
If set to true, all the `permission` related operations are skipped in module code.

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

s3_object

##### ADDITIONAL INFORMATION

AWS added the option to create S3 bucket with `ACL` `Disabled` in [Nov 2021](https://aws.amazon.com/about-aws/whats-new/2021/11/amazon-s3-object-ownership-simplify-access-management-data-s3/)) and made it the [default/suggested setting when creating S3 bucket through AWS Portal](https://docs.aws.amazon.com/AmazonS3/latest/userguide/about-object-ownership.html) .

Currently the ACL "permission" parameter defaults to `["private"]` and there is no way to tell the `s3_object` module to omit the ACL setting while uploading files to a bucket which has `set the ACL` as `disabled`.

Tried looking for a way to get the existing bucket info to determine if 'ACL' is `enabled`/`disabled`, but was not able to find what I was looking for in [API documentation](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html).

### CODE REVIEW

**Comment by mandar242** [comment]
Not sure if this is the optimal approach to move forward with this issue, keeping the PR in WIP for now to get more thoughts on it.

**Comment by gravesm** [comment]
https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client.get_bucket_ownership_controls should tell you if ACL is disabled.

**Comment by mandar242** [comment]
recheck

**Review by alinabuzachis — COMMENTED** [review]
A changelog fragment and some integration tests are needed.

**Review by jillr — CHANGES_REQUESTED** [review]
If acl_disabled isn't a user input any more it shouldn't be assigned to a module.param.  I think I understand why you did it that way, because the ACL logic is currently split up and that value needs to be accessed in two different places.  You could make `acl_disabled` a var that gets passed as an argument to `upload_s3file()` instead to deal with that.

I think it would be better in the long term though to take all the ACL logic (from `upload_s3file()` and from the series of `if` statements in `main()`) and make a new function for processing ACLs that gets called from within `upload_s3file()`.  That way if we need to add more logic related to ACLs later it will be easier to extend or modify a single function.

**Comment by mandar242** [comment]
> If acl_disabled isn't a user input any more it shouldn't be assigned to a module.param. I think I understand why you did it that way, because the ACL logic is currently split up and that value needs to be accessed in two different places. You could make `acl_disabled` a var that gets passed as an argument to `upload_s3file()` instead to deal with that.

Removed `acl_disabled` from `module.param`.

> I think it would be better in the long term though to take all the ACL logic (from `upload_s3file()` and from the series of `if` statements in `main()`) and make a new function for processing ACLs that gets called from within `upload_s3file()`. That way if we need to add more logic related to ACLs later it will be easier to extend or modify a single function.

Yes, it totally makes sense. I can work on another PR to handle ACLs logic once this one is merged?

**Review by jillr — APPROVED** [review]

File: tests/integration/targets/s3_object/tasks/copy_object_acl_disabled_bucket.yml [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+      delay: 3
+      ignore_errors: true
+
+    - name: Delete bucket created in this test
+      s3_bucket:
+        name: "{{ bucket_name }}-acl-disabled"
+        object_ownership: BucketOwnerEnforced
+        state: absent
+      register: delete_result
+
```
Comment (alinabuzachis) [inline tests/integration/targets/s3_object/tasks/copy_object_acl_disabled_bucket.yml:112]: I guess you should also remove the temp dir you created.

Comment (tremble) [reply tests/integration/targets/s3_object/tasks/copy_object_acl_disabled_bucket.yml:112]: Or better, use the helper role which adds a handler - https://github.com/ansible-collections/community.aws/blob/main/tests/integration/targets/acm_certificate/meta/main.yml

Comment (mandar242) [reply tests/integration/targets/s3_object/tasks/copy_object_acl_disabled_bucket.yml:106]: fixed.

File: tests/integration/targets/s3_object/tasks/main.yml [hunk 2]
```diff
@@ -1003,7 +1004,13 @@
...
       file:
         state: absent
-        path: "{{ tmpdir.path }}"
+        path: "{{ remote_tmp_dir }}"
+      ignore_errors: true
+
+    - name: deleting local directory with test artefacts
+      file:
+        path: '{{ remote_tmp_dir }}'
+        state: absent
```
Comment (tremble) [inline tests/integration/targets/s3_object/tasks/main.yml:1013]:
```suggestion
```

Not necessary, there's already a handler which will clean this up.

**Review by alinabuzachis — APPROVED** [review]

**Review by tremble — APPROVED** [review]

**Pull request merged**